Only keep Header values from the Primary that are not present in extension.

MSeifert04 opened the pull request:
In case a CCDData is read from a file and the primary HDU does not contain an image it will search for the first extension with data. In that case the Header is combined, but previously useful values like NAXIS, etc. were discarded in that operation.

pllim commented:
Do you check for `INHERIT=T` before actually inheriting primary header keywords? https://fits.gsfc.nasa.gov/registry/inherit.html

MSeifert04 commented:
No 😓 

That PR just makes sure the primary header values have no precedence over the extension header values (now it's the other way around). I think this is now an "always `INHERIT=T`.

pllim commented:
I haven't see any actual file with `INHERIT=F` either but it is always a possibility. I guess that is a separate issue from this PR though.

MSeifert04 commented:
I created an issue for it. It also affects other paths of the `ccddata_fits_reader` function and could break backwards-compatibility so I won't deal with it here. 

This PR is somewhat needed to fix the ccdproc test suite (which fails against master) so I would like this to be part of 2.0.2 :)

MSeifert04 commented:
@mwcraig did you have time to review this?

MSeifert04 commented:
@bsipocz This really should be in 2.0.2 because `ccdproc` kind of relies on it.

bsipocz commented:
@MSeifert04 - I'm afraid this also missed the release, but we probably can have a 2.0.3 soon after the CC meeting.

MSeifert04 commented:
okay, let me know as soon as there is a 2.0.3 section in the changelog. I'll move the changelog then. :)

bsipocz commented:
I'm waiting for the wheel builds to kick off, so hopefully within an our can push up the new changelog, etc.

bsipocz commented:
@MSeifert04 - the changelog section is now in master.

MSeifert04 commented:
The test failures seem "somewhat real" but unrelated to this PR:

<details>
=================================== FAILURES ===================================

________________________ test_skip_remote_data_astropy _________________________

pytestconfig = <_pytest.config.Config object at 0x7fc02b39c550>

    @remote_data(source='astropy')

    def test_skip_remote_data_astropy(pytestconfig):

    

        # astropy.test() has remote_data=none but we still got here somehow,

        # so fail with a helpful message

    

        if pytestconfig.getoption('remote_data') == 'none':

            pytest.fail('@remote_data was not skipped with remote_data=none')

    

        # Test Astropy URL

>       get_pkg_data_filename('galactic_center/gc_2mass_k.fits')

astropy/tests/tests/test_skip_remote_data.py:41: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

astropy/utils/data.py:627: in get_pkg_data_filename

    timeout=remote_timeout)

astropy/utils/data.py:1059: in download_file

    remote_url, timeout=timeout)) as remote:

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:154: in urlopen

    return opener.open(url, data, timeout)

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:429: in open

    response = self._open(req, data)

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:447: in _open

    '_open', req)

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:407: in _call_chain

    result = func(*args)

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:1228: in http_open

    return self.do_open(httplib.HTTPConnection, req)

/home/travis/miniconda/envs/test/lib/python2.7/urllib2.py:1195: in do_open

    h.request(req.get_method(), req.get_selector(), req.data, headers)

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:1042: in request

    self._send_request(method, url, body, headers)

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:1082: in _send_request

    self.endheaders(body)

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:1038: in endheaders

    self._send_output(message_body)

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:882: in _send_output

    self.send(msg)

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:844: in send

    self.connect()

/home/travis/miniconda/envs/test/lib/python2.7/httplib.py:821: in connect

    self.timeout, self.source_address)

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (('astropy.org', 80), 10.0, None), kwargs = {}, addr_arg = 0

valid_hosts = ('localhost', '127.0.0.1', 'data.astropy.org', '162.255.119.248', 'astropy.stsci.edu', '130.167.180.4')

valid_host = 'astropy.stsci.edu', valid_host_ip = '130.167.180.4'

hostname = 'testing-docker-bd5e3ac0-6619-49e6-ad18-4ff6f74fea0d'

fqdn = 'testing-docker-bd5e3ac0-6619-49e6-ad18-4ff6f74fea0d'

    def new_function(*args, **kwargs):

        if isinstance(args[0], socket.socket):

            if not args[0].family in (socket.AF_INET, socket.AF_INET6):

                # Should be fine in all but some very obscure cases

                # More to the point, we don't want to affect AF_UNIX

                # sockets.

                return original_function(*args, **kwargs)

            host = args[1][0]

            addr_arg = 1

            valid_hosts = ('localhost', '127.0.0.1', '::1')

        else:

            # The only other function this is used to wrap currently is

            # socket.create_connection, which should be passed a 2-tuple, but

            # we'll check just in case

            if not (isinstance(args[0], tuple) and len(args[0]) == 2):

                return original_function(*args, **kwargs)

    

            host = args[0][0]

            addr_arg = 0

            valid_hosts = ('localhost', '127.0.0.1')

    

        if allow_astropy_data:

            for valid_host in ('data.astropy.org', 'astropy.stsci.edu'):

                valid_host_ip = socket.gethostbyname(valid_host)

                valid_hosts += (valid_host, valid_host_ip)

    

        hostname = socket.gethostname()

        fqdn = socket.getfqdn()

    

        if host in (hostname, fqdn):

            host = 'localhost'

            new_addr = (host, args[addr_arg][1])

            args = args[:addr_arg] + (new_addr,) + args[addr_arg + 1:]

    

        if any(h in host for h in valid_hosts):

            return original_function(*args, **kwargs)

        else:

            raise IOError("An attempt was made to connect to the internet "

                          "by a test that was not marked `remote_data`. The "

>                         "requested host was: {0}".format(host))

E           IOError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: astropy.org

astropy/tests/disable_internet.py:75: IOError

__________________ TestBasic.test_tick_angles_non_square_axes __________________

args = (<astropy.visualization.wcsaxes.tests.test_images.TestBasic object at 0x7fbfdf48e450>,)

kwargs = {}

baseline_dir = 'http://data.astropy.org/testing/astropy/2017-07-12T14:12:26.217559/1.5.x/'

baseline_remote = True

inspect = <module 'inspect' from '/home/travis/miniconda/envs/test/lib/python2.7/inspect.pyc'>

fig = <matplotlib.figure.Figure object at 0x7fbfdf48e3d0>

filename = 'tick_angles_non_square_axes.png', result_dir = '/tmp/tmpgXiYme'

test_image = '/tmp/tmpgXiYme/tick_angles_non_square_axes.png'

baseline_image_ref = '/tmp/tmp2AqF80/downloaded'

baseline_image = '/tmp/tmpgXiYme/baseline-tick_angles_non_square_axes.png'

    @wraps(item.function)

    def item_function_wrapper(*args, **kwargs):

    

        baseline_dir = compare.kwargs.get('baseline_dir', None)

        if baseline_dir is None:

            if self.baseline_dir is None:

                baseline_dir = os.path.join(os.path.dirname(item.fspath.strpath), 'baseline')

            else:

                baseline_dir = self.baseline_dir

        else:

            if not baseline_dir.startswith(('http://', 'https://')):

                baseline_dir = os.path.join(os.path.dirname(item.fspath.strpath), baseline_dir)

    

        baseline_remote = baseline_dir.startswith('http')

    

        with plt.style.context(style), switch_backend(backend):

    

            # Run test and get figure object

            import inspect

            if inspect.ismethod(original):  # method

                # In some cases, for example if setup_method is used,

                # original appears to belong to an instance of the test

                # class that is not the same as args[0], and args[0] is the

                # one that has the correct attributes set up from setup_method

                # so we ignore original.__self__ and use args[0] instead.

                fig = original.__func__(*args, **kwargs)

            else:  # function

                fig = original(*args, **kwargs)

    

            if remove_text:

                MplImageComparisonTest.remove_text(fig)

    

            # Find test name to use as plot name

            filename = compare.kwargs.get('filename', None)

            if filename is None:

                filename = item.name + '.png'

                filename = filename.replace('[', '_').replace(']', '_')

                filename = filename.replace('/', '_')

                filename = filename.replace('_.png', '.png')

    

            # What we do now depends on whether we are generating the

            # reference images or simply running the test.

            if self.generate_dir is None:

    

                # Save the figure

                result_dir = tempfile.mkdtemp(dir=self.results_dir)

                test_image = os.path.abspath(os.path.join(result_dir, filename))

    

                fig.savefig(test_image, **savefig_kwargs)

                plt.close(fig)

    

                # Find path to baseline image

                if baseline_remote:

                    baseline_image_ref = _download_file(baseline_dir + filename)

                else:

                    baseline_image_ref = os.path.abspath(os.path.join(os.path.dirname(item.fspath.strpath), baseline_dir, filename))

    

                if not os.path.exists(baseline_image_ref):

                    pytest.fail("Image file not found for comparison test. "

                                "(This is expected for new tests.)\nGenerated Image: "

                                "\n\t{test}".format(test=test_image), pytrace=False)

    

                # distutils may put the baseline images in non-accessible places,

                # copy to our tmpdir to be sure to keep them in case of failure

                baseline_image = os.path.abspath(os.path.join(result_dir, 'baseline-' + filename))

                shutil.copyfile(baseline_image_ref, baseline_image)

    

>               msg = compare_images(baseline_image, test_image, tol=tolerance)

/home/travis/miniconda/envs/test/lib/python2.7/site-packages/pytest_mpl/plugin.py:218: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = '/tmp/tmpgXiYme/baseline-tick_angles_non_square_axes.png'

actual = '/tmp/tmpgXiYme/tick_angles_non_square_axes.png', tol = 1.5

in_decorator = False

    def compare_images(expected, actual, tol, in_decorator=False):

        """

        Compare two "image" files checking differences within a tolerance.

    

        The two given filenames may point to files which are convertible to

        PNG via the `.converter` dictionary. The underlying RMS is calculated

        with the `.calculate_rms` function.

    

        Parameters

        ----------

        expected : str

            The filename of the expected image.

        actual :str

            The filename of the actual image.

        tol : float

            The tolerance (a color value difference, where 255 is the

            maximal difference).  The test fails if the average pixel

            difference is greater than this value.

        in_decorator : bool

            If called from image_comparison decorator, this should be

            True. (default=False)

    

        Example

        -------

        img1 = "./baseline/plot.png"

        img2 = "./output/plot.png"

        compare_images( img1, img2, 0.001 ):

    

        """

        if not os.path.exists(actual):

            msg = "Output image %s does not exist." % actual

            raise Exception(msg)

    

        if os.stat(actual).st_size == 0:

            msg = "Output image file %s is empty." % actual

            raise Exception(msg)

    

        verify(actual)

    

        # Convert the image to png

        extension = expected.split('.')[-1]

    

        if not os.path.exists(expected):

            raise IOError('Baseline image %r does not exist.' % expected)

    

        if extension != 'png':

            actual = convert(actual, False)

            expected = convert(expected, True)

    

        # open the image files and remove the alpha channel (if it exists)

>       expectedImage = _png.read_png_int(expected)

E       ValueError: invalid PNG header

/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/testing/compare.py:313: ValueError

_______________________ TestBasic.test_ticks_regression ________________________

args = (<astropy.visualization.wcsaxes.tests.test_images.TestBasic object at 0x7fc0107855d0>,)

kwargs = {}

baseline_dir = 'http://data.astropy.org/testing/astropy/2017-07-12T14:12:26.217559/1.5.x/'

baseline_remote = True

inspect = <module 'inspect' from '/home/travis/miniconda/envs/test/lib/python2.7/inspect.pyc'>

fig = <matplotlib.figure.Figure object at 0x7fbfdf331c90>

filename = 'test_ticks_regression_1.png', result_dir = '/tmp/tmpNLOULj'

test_image = '/tmp/tmpNLOULj/test_ticks_regression_1.png'

baseline_image_ref = '/tmp/tmpKxEBSn/downloaded'

baseline_image = '/tmp/tmpNLOULj/baseline-test_ticks_regression_1.png'

    @wraps(item.function)

    def item_function_wrapper(*args, **kwargs):

    

        baseline_dir = compare.kwargs.get('baseline_dir', None)

        if baseline_dir is None:

            if self.baseline_dir is None:

                baseline_dir = os.path.join(os.path.dirname(item.fspath.strpath), 'baseline')

            else:

                baseline_dir = self.baseline_dir

        else:

            if not baseline_dir.startswith(('http://', 'https://')):

                baseline_dir = os.path.join(os.path.dirname(item.fspath.strpath), baseline_dir)

    

        baseline_remote = baseline_dir.startswith('http')

    

        with plt.style.context(style), switch_backend(backend):

    

            # Run test and get figure object

            import inspect

            if inspect.ismethod(original):  # method

                # In some cases, for example if setup_method is used,

                # original appears to belong to an instance of the test

                # class that is not the same as args[0], and args[0] is the

                # one that has the correct attributes set up from setup_method

                # so we ignore original.__self__ and use args[0] instead.

                fig = original.__func__(*args, **kwargs)

            else:  # function

                fig = original(*args, **kwargs)

    

            if remove_text:

                MplImageComparisonTest.remove_text(fig)

    

            # Find test name to use as plot name

            filename = compare.kwargs.get('filename', None)

            if filename is None:

                filename = item.name + '.png'

                filename = filename.replace('[', '_').replace(']', '_')

                filename = filename.replace('/', '_')

                filename = filename.replace('_.png', '.png')

    

            # What we do now depends on whether we are generating the

            # reference images or simply running the test.

            if self.generate_dir is None:

    

                # Save the figure

                result_dir = tempfile.mkdtemp(dir=self.results_dir)

                test_image = os.path.abspath(os.path.join(result_dir, filename))

    

                fig.savefig(test_image, **savefig_kwargs)

                plt.close(fig)

    

                # Find path to baseline image

                if baseline_remote:

                    baseline_image_ref = _download_file(baseline_dir + filename)

                else:

                    baseline_image_ref = os.path.abspath(os.path.join(os.path.dirname(item.fspath.strpath), baseline_dir, filename))

    

                if not os.path.exists(baseline_image_ref):

                    pytest.fail("Image file not found for comparison test. "

                                "(This is expected for new tests.)\nGenerated Image: "

                                "\n\t{test}".format(test=test_image), pytrace=False)

    

                # distutils may put the baseline images in non-accessible places,

                # copy to our tmpdir to be sure to keep them in case of failure

                baseline_image = os.path.abspath(os.path.join(result_dir, 'baseline-' + filename))

                shutil.copyfile(baseline_image_ref, baseline_image)

    

>               msg = compare_images(baseline_image, test_image, tol=tolerance)

/home/travis/miniconda/envs/test/lib/python2.7/site-packages/pytest_mpl/plugin.py:218: 

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected = '/tmp/tmpNLOULj/baseline-test_ticks_regression_1.png'

actual = '/tmp/tmpNLOULj/test_ticks_regression_1.png', tol = 1.5

in_decorator = False

    def compare_images(expected, actual, tol, in_decorator=False):

        """

        Compare two "image" files checking differences within a tolerance.

    

        The two given filenames may point to files which are convertible to

        PNG via the `.converter` dictionary. The underlying RMS is calculated

        with the `.calculate_rms` function.

    

        Parameters

        ----------

        expected : str

            The filename of the expected image.

        actual :str

            The filename of the actual image.

        tol : float

            The tolerance (a color value difference, where 255 is the

            maximal difference).  The test fails if the average pixel

            difference is greater than this value.

        in_decorator : bool

            If called from image_comparison decorator, this should be

            True. (default=False)

    

        Example

        -------

        img1 = "./baseline/plot.png"

        img2 = "./output/plot.png"

        compare_images( img1, img2, 0.001 ):

    

        """

        if not os.path.exists(actual):

            msg = "Output image %s does not exist." % actual

            raise Exception(msg)

    

        if os.stat(actual).st_size == 0:

            msg = "Output image file %s is empty." % actual

            raise Exception(msg)

    

        verify(actual)

    

        # Convert the image to png

        extension = expected.split('.')[-1]

    

        if not os.path.exists(expected):

            raise IOError('Baseline image %r does not exist.' % expected)

    

        if extension != 'png':

            actual = convert(actual, False)

            expected = convert(expected, True)

    

        # open the image files and remove the alpha channel (if it exists)

>       expectedImage = _png.read_png_int(expected)

E       ValueError: invalid PNG header

/home/travis/miniconda/envs/test/lib/python2.7/site-packages/matplotlib/testing/compare.py:313: ValueError

=============================== warnings summary ===============================
</details>

MSeifert04 commented:
@mwcraig @crawfordsm Could you have a look? We merged a similar PR in `ccdproc` already.

pllim commented:
There are some conflicts -- FYI

MSeifert04 commented:
Thanks for the heads up, conflicts should be fixed now :)

MSeifert04 commented:
Test failure is unrelated (remote tests)...
